### PR TITLE
Fix Thunderloop on master + reduce REDIS gets to improve speed

### PR DESCRIPTION
**Pull request by itsarune**
### Description

This PR gets master's thunderloop to be clean once again. It does the following:

- fixes a segfault when network_service_ isn't initialized on the first loop
- only grab from the redis store on initialization, rather than every time while looping (this should help improve thunderloop's control loop speed)

### Testing Done

~~pending on a robot~~
tested on Aimbot (thunderloop flashing + robot diagnostics primitives)

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

**Review by LiCody — COMMENTED** [review]
Theres a few more redis calls in thunderloop for kick slope/constant and chip constant do we want to leave those in? I think it comes down to whether we want to use a timeouts to restart thunderloop for a robot that kicked too hard vs changing it on the fly. We already aim for 6 m/s vs the SSL limit of 6.5 m/s so I think we can move those redis calls out as well

**Review by itsarune — COMMENTED** [review]
the other redis calls (kick_slope, kick_constant, chip_pulse_width) probably shouldn't be necessary. It seems that after calibration, we rarely change these values

imo if we want to make on the fly changes, we should be using dynamic parameters (which we can do for controlling min, max pass speed)

**Comment by nimazareian** [comment]
> imo if we want to make on the fly changes, we should be using dynamic parameters (which we can do for controlling min, max pass speed)

That'd be great, though dynamic parameters are currently not sent to the robots if they are changed on the fly, so we would need to implement that.

File: src/software/jetson_nano/thunderloop.cpp [hunk 1]
```diff
@@ -162,26 +153,6 @@ void Thunderloop::runLoop()
...
-                // Log milliseconds since last world received if we are timing out
-                int milliseconds_elapsed_since_last_world = static_cast<int>(
-                    static_cast<double>(world_result.tv_sec) * MILLISECONDS_PER_SECOND +
-                    static_cast<double>(world_result.tv_nsec) *
-                        MILLISECONDS_PER_NANOSECOND);
-
-                LOG(INFO) << "World timeout, overriding with StopPrimitive\n"
-                          << "Milliseconds since last world: "
-                          << milliseconds_elapsed_since_last_world;
-            }
```
Comment (itsarune) [inline src/software/jetson_nano/thunderloop.cpp]: @AdenC123 I didn't realize this earlier but when we send RobotDiagnostics primitives, we don't send Worlds, so robot diagnostics primitives gets overridden by this override.

Do you think we should find a workaround for Worlds @nimazareian? or should we only override expired PrimitiveSets?

**Review by nimazareian — DISMISSED** [review]
LGTM! :100:

File: src/software/jetson_nano/thunderloop.cpp (hunk 1 above)
Comment (nimazareian) [reply src/software/jetson_nano/thunderloop.cpp]: That's interesting. I think it'll be helpful to still have this check, since if vision for whatever reason suddenly dies, we don't want the robots to go crazy. The only way that comes to my mind right now for fixing this would be to have a flag in the Primitive proto which specifies whether it is sent from AI or from diagnostics, and only having the above logic work for when it is from AI.

**Review by LiCody — DISMISSED** [review]
LGTM

File: src/software/jetson_nano/redis/redis_client_test.cpp [hunk 2]
```diff
@@ -9,6 +9,9 @@
 #include "cpp_redis/core/subscriber.hpp"
 #include "software/test_util/test_util.h"
 
+// TODO #2805: Convert this into a test fixture so that we can start and run a REDIS
+// server during the lifetime of the test fixture. Then re-enable all tests
```
Comment (nimazareian) [inline src/software/jetson_nano/redis/redis_client_test.cpp:13]: Is there a reason why they can't remain enabled until this task is done?

Comment (itsarune) [reply src/software/jetson_nano/redis/redis_client_test.cpp:13]: because we need to have a redis server up and running for this to work

Comment (itsarune) [reply src/software/jetson_nano/redis/redis_client_test.cpp:13]: Because these tests assume that there's a redis server running to GET and SET on

Comment (itsarune) [reply src/software/jetson_nano/redis/redis_client_test.cpp:13]: It's doable for me to set this up via a test fixture but I figured it wasn't a high priority thing and we can do it later

Comment (nimazareian) [reply src/software/jetson_nano/redis/redis_client_test.cpp:13]: Okay sounds good

**Review by nimazareian — COMMENTED** [review]
LGTM, pending the merge conflicts

File: src/software/jetson_nano/redis/redis_client.cpp [hunk 3]
```diff
@@ -45,16 +45,50 @@ RedisClient::RedisClient(std::string host, size_t port)
...
-std::string RedisClient::get(const std::string &key)
+std::string RedisClient::getSync(const std::string &key)
 {
     auto future = client_.get(key);
     client_.commit();
     return future.get().as_string();
 }
 
-void RedisClient::set(const std::string &key, const std::string &value)
+void RedisClient::getAsyncNoCommit(const std::string &key,
```
Comment (Bvasilchikov) [inline src/software/jetson_nano/redis/redis_client.cpp:55]: do we need this function? With no commit the caller is relying on other functions to be called that use commit. Might become buggy with the callback in the future.

**Review by Bvasilchikov — DISMISSED** [review]
everything lgtm, just one comment on the api

File: src/software/jetson_nano/redis/redis_client.cpp (hunk 3 above)
Comment (itsarune) [reply src/software/jetson_nano/redis/redis_client.cpp:55]: I guess what I wanted to do was reduce the amount of "network transfers" we have to make to the redis server (even though its running on the nano) I figured having more control over how we commit will let us send multiple in one packet

**Review by nimazareian — DISMISSED** [review]
Just tested it out and it works with diagnostics!

**Review by nimazareian — DISMISSED** [review]

File: src/software/jetson_nano/thunderloop.cpp [hunk 4]
```diff
@@ -21,31 +21,49 @@ extern int clock_nanosleep(clockid_t __clock_id, int __flags,
...
+    LOG(INFO)
+        << "THUNDERLOOP: to update Thunderloop configuration, change REDIS store and restart Thunderloop";
 }
 
 Thunderloop::~Thunderloop() {}
 
 /*
  * Run the main robot loop!
  */
-[[noreturn]] void Thunderloop::runLoop()
```
Comment (LiCody) [inline src/software/jetson_nano/thunderloop.cpp]: is this change intentional?

Comment (itsarune) [reply src/software/jetson_nano/thunderloop.cpp]: not intentional, updated

**Review by nimazareian — APPROVED** [review]

**Review by LiCody — APPROVED** [review]
